src/libs/kata-sys-util/src/mount.rs

```diff
@@ -0,0 +1,1087 @@
+    #[error("Can not read metadata of {0}, {1}")]
+    ReadMetadata(PathBuf, io::Error),
+    #[error("Can not remount {0}: {1}")]
+    Remount(PathBuf, nix::Error),
+    #[error("Can not find mountpoint for {0}")]
+    NoMountEntry(String),
+    #[error("Can not umount {0}, {1}")]
+    Umount(PathBuf, io::Error),
+}
+
+/// A specialized version of `std::result::Result` for mount operations.
+pub type Result<T> = std::result::Result<T, Error>;
+
+/// Information of mount record from `/proc/mounts`.
+pub struct LinuxMountInfo {
+    /// Source of mount, first field of records from `/proc/mounts`.
+    pub device: String,
+    /// Destination of mount, second field of records from `/proc/mounts`.
+    pub path: String,
+    /// Filesystem type of mount, third field of records from `/proc/mounts`.
+    pub fs_type: String,
+}
+
+/// Get the device and file system type of a mount point by parsing `/proc/mounts`.
+pub fn get_linux_mount_info(mount_point: &str) -> Result<LinuxMountInfo> {
+    let mount_file = fs::File::open(PROC_MOUNTS_FILE)?;
+    let lines = io::BufReader::new(mount_file).lines();
+
+    for mount in lines.flatten() {
+        let fields: Vec<&str> = mount.split(' ').collect();
+
+        if fields.len() != PROC_FIELDS_PER_LINE {
+            return Err(Error::InvalidMountEntry(
+                PROC_FIELDS_PER_LINE,
+                fields.len(),
+                mount,
+            ));
+        }
+
+        if mount_point == fields[PROC_PATH_INDEX] {
+            return Ok(LinuxMountInfo {
+                device: fields[PROC_DEVICE_INDEX].to_string(),
+                path: fields[PROC_PATH_INDEX].to_string(),
+                fs_type: fields[PROC_TYPE_INDEX].to_string(),
+            });
+        }
+    }
+
+    Err(Error::NoMountEntry(mount_point.to_owned()))
+}
+
+/// Recursively create destination for a mount.
+///
+/// For a normal mount, the destination will always be a directory. For bind mount, the destination
+/// must be a directory if the source is a directory, otherwise the destination must be a normal
+/// file. If directories are created, their permissions are initialized to MountPerm.
+///
+/// # Safety
+///
+/// Every container has a root filesystems `rootfs`. When creating bind mounts for a container,
+/// the destination should always be within the container's `rootfs`. Otherwise it's a serious
+/// security flaw for container to read/override host side filesystem contents. Please refer to
+/// following CVEs for example:
+/// - [CVE-2021-30465](https://github.com/opencontainers/runc/security/advisories/GHSA-c3xm-pvg7-gh7r)
+///
+/// To ensure security, the `create_mount_destination()` function takes an extra parameter `root`,
+/// which is used to ensure that `dst` is within the specified directory. And a safe version of
+/// `PathBuf` is returned to avoid TOCTTOU type of flaws.
+pub fn create_mount_destination<S: AsRef<Path>, D: AsRef<Path>, R: AsRef<Path>>(
+    src: S,
+    dst: D,
+    _root: R,
```

Comment:
Is this going to be enabled when https://github.com/kata-containers/kata-containers/pull/3462 lands?

Reply:
Yes, there are several TODOs in mount.rs, which will be handled after merging #3462.